hashing/src/chunk_with_proof.rs

```diff
@@ -0,0 +1,311 @@
+pub struct ChunkWithProof {
+    proof: IndexedMerkleProof,
+    chunk: Vec<u8>,
+}
+
+impl ToBytes for ChunkWithProof {
+    fn to_bytes(&self) -> Result<Vec<u8>, casper_types::bytesrepr::Error> {
+        let mut result = allocate_buffer(self)?;
+        result.append(&mut self.proof.to_bytes()?);
+        result.append(&mut Bytes::from(self.chunk.as_slice()).to_bytes()?);
+        Ok(result)
+    }
+
+    fn serialized_length(&self) -> usize {
+        self.proof.serialized_length() + self.chunk.serialized_length()
+    }
+}
+
+impl FromBytes for ChunkWithProof {
+    fn from_bytes(bytes: &[u8]) -> Result<(Self, &[u8]), casper_types::bytesrepr::Error> {
+        let ((proof, chunk), remainder) = <(IndexedMerkleProof, Bytes)>::from_bytes(bytes)?;
+
+        Ok((
+            ChunkWithProof::try_from(ChunkWithProofDeserializeValidator {
+                proof,
+                chunk: chunk.into(),
+            })
+            .map_err(|_| casper_types::bytesrepr::Error::Validation)?,
+            remainder,
+        ))
+    }
+}
+
+impl ChunkWithProof {
+    #[cfg(test)]
+    pub const CHUNK_SIZE: usize = 1;
+
+    // TODO[RC]: Uncomment when the `ChunkWithProof` is actually used in the code
+    // #[cfg(not(test))]
+    // pub const CHUNK_SIZE: usize = 1_048_576; // 2^20
+
+    #[cfg(test)]
+    pub fn new(data: &[u8], index: u64) -> Result<Self, error::MerkleConstructionError> {
+        if data.len() < Self::CHUNK_SIZE * (index as usize) {
+            return Err(error::MerkleConstructionError::IndexOutOfBounds {
+                count: data.chunks(Self::CHUNK_SIZE).len() as u64,
+                index,
+            });
+        }
+
+        let (proof, chunk) = if data.is_empty() {
+            (IndexedMerkleProof::new([Digest::hash(&[])], index)?, vec![])
+        } else {
+            (
+                IndexedMerkleProof::new(data.chunks(Self::CHUNK_SIZE).map(Digest::hash), index)?,
+                data[Self::CHUNK_SIZE * (index as usize)
+                    ..data.len().min(Self::CHUNK_SIZE * ((index as usize) + 1))]
+                    .to_vec(),
+            )
+        };
+
+        Ok(ChunkWithProof { proof, chunk })
+    }
+
+    fn is_valid(&self) -> bool {
+        let chunk_hash = Digest::hash(self.chunk());
+        self.proof
+            .merkle_proof()
+            .first()
+            .map_or(false, |first_hash| chunk_hash == *first_hash)
+    }
+
+    /// Get a reference to the chunk with proof's chunk.
+    pub fn chunk(&self) -> &[u8] {
+        self.chunk.as_slice()
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use crate::{error, Digest};
+    use casper_types::bytesrepr::{FromBytes, ToBytes};
+    use proptest::proptest;
+    use rand::Rng;
+    use std::{convert::TryInto, ops::Range};
+
+    use crate::chunk_with_proof::ChunkWithProof;
+
+    fn testing_data_size_range(minimum_size: usize) -> Range<usize> {
+        minimum_size..512usize
+    }
```

Comment:
Instead of this, maybe try:

```rust
struct TestDataSize(u16);

impl proptest::prelude::Arbitrary for TestDataSize {
    type Parameters = ();
    type Strategy = BoxedStrategy<Self>;

    fn arbitrary_with(_: Self::Parameters) -> Self::Strategy {
        (minimum_size..512usize).prop_map(TestDataSize)
    }
}
```

Then below you can ditch the `proptest!` macro and use `use proptest_attr_macro::proptest;` instead.

Reply:
I ditched the `testing_data_size_range()` function in favor of explicit ranges, like so: `fn generates_valid_proof(chunk_count in 0usize..32usize)`. Is that ok?

Reply:
As I've mentioned below, I still think we should avoid the `proptest!` macro and use the `#[proptest]` attr macro instead.

Reply:
I ended up with two different structs:
* `TestDataSize` - for range **0**..32
* `TestDataSizeAtLeastTwoChunks` - for range **2**..32

Normally, I would use a hand-crafted strategy, like so:

```
fn my_strategy() -> Impl<Strategy = BoxedStrategy<TestDataSize>> {
    any::<TestDataSize>().prop_map(|_| TestDataSize(2)) // Pass minimum value
}
```

but, on the contrary to raw `proptest!` macro, the procedural variant `#[proptest]` doesn't allow the `in` syntax, so I can't write:
`fn generates_valid_proof(test_data in my_strategy)`
and the following, obviously, doesn't compile:
`fn generates_valid_proof(test_data: my_strategy())`

Since I can't see a clear solution to that, I think the benefits of `#[proptest]` are not worth complicating the code.

Reply:
>  I still think we should avoid the proptest! macro and use the #[proptest] attr macro instead.

Why @xcthulhu ?

Reply:
The procedural macro plays better with `rutstfmt` and `clippy`.  Thanks for doing this @rafal-ch !